### PR TITLE
fix: finetune model inference needs the dtype fix to work with flash-attn

### PR DESCRIPTION
**Pull request by maximegmd**

### CODE REVIEW

File: src/axolotl/utils/models.py [hunk 1]
```diff
@@ -363,16 +364,19 @@ def load_model(
...
+        fix_dtype = True
 
-        # LlamaRMSNorm layers are in fp32 after kbit_training, so we need to
-        # convert them back to fp16/bf16 for flash-attn compatibility.
-        if cfg.flash_attention and cfg.is_llama_derived_model:
-            for name, module in model.named_modules():
-                if "norm" in name:
+    # LlamaRMSNorm layers are in fp32 after kbit_training or full finetune, so we need to
+    # convert them back to fp16/bf16 for flash-attn compatibility.
+    if (fix_dtype or cfg.adapter == "" or cfg.adapter is None) and (
```
Comment (winglian) [inline src/axolotl/utils/models.py]: `if (fix_dtype or cfg.adapter == "" or cfg.adapter is None) and (`
could be simplified to 
`if (fix_dtype or not cfg.adapter) and (`

**Review by winglian — APPROVED** [review]
lgtm

File: src/axolotl/utils/models.py [hunk 2]
```diff
@@ -355,6 +355,7 @@ def load_model(
             if hasattr(module, "weight"):
                 module.to(torch.float32)
 
+    fix_dtype = False
```
Comment (tmm1) [inline src/axolotl/utils/models.py]:
```suggestion
    fix_dtype = not cfg.adapter
```

File: src/axolotl/utils/models.py [hunk 3]
```diff
@@ -363,16 +364,19 @@ def load_model(
...
+        fix_dtype = True
 
-        # LlamaRMSNorm layers are in fp32 after kbit_training, so we need to
-        # convert them back to fp16/bf16 for flash-attn compatibility.
-        if cfg.flash_attention and cfg.is_llama_derived_model:
-            for name, module in model.named_modules():
-                if "norm" in name:
+    # LlamaRMSNorm layers are in fp32 after kbit_training or full finetune, so we need to
+    # convert them back to fp16/bf16 for flash-attn compatibility.
+    if (fix_dtype or not cfg.adapter) and (
```
Comment (tmm1) [inline src/axolotl/utils/models.py]:
```suggestion
    if fix_dtype and (
```

File: src/axolotl/utils/models.py (hunk 2 above)
Comment (tmm1) [reply src/axolotl/utils/models.py]: let's also rename this, something like need_fa2_dtype

**Pull request merged**